### PR TITLE
Replace the blib2to3 tokenizer with pytokens

### PR DESCRIPTION
**Pull request by tusharsadhwani**
### Description

Replaces black's tokenizer with a from-scratch rewrite.

Should help resolve problems like #4520 and others.

### What's left

- `ERRORTOKEN` handling is not very proper just yet, it'll require some tweaks.
- Edge case tests: Currently 366/381 tests are passing.

### CODE REVIEW

**Comment by tusharsadhwani** [comment]
@JelleZijlstra with this, the test suite is fully passing. Primer is failing (mostly just because some file in `hypothesis` failed to parse), I'll be working on that, but the code should be good for a first review.

**Comment by MeGaGiGaGon** [comment]
I don't think we currently have any tests for this, but I just linked the above two issues here because they are the same bug in the parser where `\r`s cause issues. The most minimal reproduction is `{\r}`, and since this is a parser rewrite hopefully it can be solved. Note that this is currently only observable by directly calling internal methods due to how black reads input.

**Comment by tusharsadhwani** [comment]
Thanks for linking this, I'll make sure these parse identically to how CPython does it.

**Comment by tusharsadhwani** [comment]
Okay, primer is fixed, and all tests are green.

**Comment by JelleZijlstra** [comment]
Might this fuzzer failure indicate a bug?

```
Falsifying example: test_idempotent_any_syntatically_valid_python(
    src_contents='\n#\r#',
    mode=Mode(target_versions=set(),
     line_length=88,
     string_normalization=False,
     is_pyi=False,
     is_ipynb=False,
     skip_source_first_line=False,
     magic_trailing_comma=False,
     python_cell_magics=set(),
     preview=False,
     unstable=False,
     enabled_features=set()),  # or any other generated value
)
```

File: src/blib2to3/pgen2/tokenize.py [hunk 1]
```diff
@@ -45,1070 +43,218 @@
...
-    Each call to the function should return one line of input as a string.
 
-    The second parameter, tokeneater, must also be a callable object. It is
-    called once for each token, with five arguments, corresponding to the
-    tuples generated by generate_tokens().
+def transform_whitespace(
+    token: pytokens.Token, source: str, prev_token: Optional[pytokens.Token]
+) -> pytokens.Token:
+    r"""
+    Black treats `\\\n` at the end of a line as a 'NL' token, while it
```
Comment (JelleZijlstra) [inline src/blib2to3/pgen2/tokenize.py:105]: That doesn't sound particularly intentional, I'd be open to changing Black to remove this divergence.

**Comment by tusharsadhwani** [comment]
> Might this fuzzer failure indicate a bug?

Yeah, but I'm pretty sure it is a bug in CPython. For now we can work it out in the tokenizer though. I'll add a flag in `pytokens` to fix it.

File: src/blib2to3/pgen2/tokenize.py (hunk 1 above)
Comment (tusharsadhwani) [reply src/blib2to3/pgen2/tokenize.py:105]: Feel free to! I can give you a test case with the expected behaviour.

Comment (tusharsadhwani) [reply src/blib2to3/pgen2/tokenize.py:105]: So, this is enough as a test case actually:

```python
a \
  b
```

But, the reason this probably exists is to support formatting [this file](https://github.com/psf/black/blob/main/tests/data/cases/backslash_before_indent.py):

```python
class Plotter:
\
    pass

class AnotherCase:
  \
    """Some
    \
    Docstring
    """
```

**Comment by MeGaGiGaGon** [comment]
I found another issue that this should close, https://github.com/psf/black/issues/2318

**Comment by JelleZijlstra** [comment]
@MeGaGiGaGon no, that's about the parser, not the tokenizer. It would still be good long term to switch to a different parser, because the current parser only supports Python's recent grammar changes through a series of hacks.

**Comment by hauntsaninja** [comment]
This seems awesome! Could be worth doing a quick benchmark though with a mypyc build (I guess we could potentially mypyc pytokens if there is a perf difference). I think to do that you have to set these env vars https://github.com/psf/black/blob/main/pyproject.toml#L181 (but honestly not too sure how hatch-mypyc is setup)

**Comment by tusharsadhwani** [comment]
@JelleZijlstra edge cases have been taken care of.

@hauntsaninja I just tested `mypyc` on `pytokens` locally, and it got 2x faster at tokenizing the `black/profiling` folder.

File: src/blib2to3/pgen2/tokenize.py (hunk 1 above)
Comment (tusharsadhwani) [reply src/blib2to3/pgen2/tokenize.py:105]: I tried removing this edge case, but it just seems to break the backslash before indent test case. Let's just keep it as-is for now and try to fix it later?

**Comment by cobaltt7** [comment]
Resolves #4588 too

**Comment by tusharsadhwani** [comment]
@JelleZijlstra can you confirm if the PR is good to review, or are changes needed?

**Review by JelleZijlstra — APPROVED** [review]
Let's do it!

**Comment by tusharsadhwani** [comment]
@JelleZijlstra I don't have merge rights just yet :P

**Comment by JelleZijlstra** [comment]
Sorry I had to resolve a merge conflict and then wait for CI
